### DISCUSSION TITLE
Guidance for logging non-UVM classes (for example, BFMs)

### DISCUSSION THREAD

**Discussion by timothyscherer** [pyuvm/pyuvm#150] [Q&A] [answered]
I am working on a testbench which includes a few new BFMs. Of course, these BFMs do not utilize any pyuvm base classes, and as such do not contain a logger or anything to that effect.

As a general practice, they do all have a class member named `log`, which worked fine when I was only using cocotb. Now I am wondering if there is an elegant way to integrate those loggers with the pyuvm loggers. Right now, I can simply set the log members to the monitor logger, and that seems to work, but it is not as informative (as to which module it comes from). Should I just leave the loggers alone?

I am thinking about the possibility of a JSON schema where I can set the log levels throughout the hierarchy using the ConfigDB, could this be an effective approach to customizing log levels?

**Comment by raysalemi** [comment]
What if you passed `self.logger` from your `uvm_component` to the BFM? 

Unless I'm mistaken, the BFM will then send its log messages through the UVM hierarchy, and you can use the **pyuvm** functions such as `self.set_logging_level()` to set the logging level for that component.

**Reply by timothyscherer** [reply]
That was my first impression, I'm just not a fan of the way those messages would all get combined into that one component. My BFM is going to contain multiple BFMs once it is done, and I would like to be able to differentiate those messages.

Using `getChild()` seems appropriate here I think, but this uses the existing PyuvmFormatter which is "hardcoded" on creation of the parent. Right now, the log message at least shows the filename it came from to help differentiate, but it would be nice if the new path showed in the log message itself. As far as I can tell, there isn't a nice way to reconfigure the handler + formatter to append something to the path.

```py
class Agent(uvm_agent):
    def build_phase(self):
        super().build_phase()
        self.cdb_set("bus", cocotb.top, "")
        bfm = FunctionalModel(self.cdb_get("bus"))
        bfm.log = self.logger.getChild("bfm")
```

Actual result:
```log
# 4060400.00ns INFO     ..t\riox_i2c_ctrl\__init__.py(295) [uvm_test_top.env.agent]: double_write(paddr=0x10, port0=0x03, data0=0x87, port1=0x00, data1=0x36)
```

Desired result:
```log
# 4060400.00ns INFO     ..t\riox_i2c_ctrl\__init__.py(295) [uvm_test_top.env.agent.bfm]: double_write(paddr=0x10, port0=0x03, data0=0x87, port1=0x00, data1=0x36)
```

**Comment by raysalemi** [comment, marked as answer]
You can change the formatter on a component. There is a chapter in the book
on this.

I find the *logging* module to be super complicated and non-intuitive, so I
tried to encapsulate its behavior into *pyuvm* because I didn't want to
think about it.
